segmentation.py

```diff
@@ -0,0 +1,575 @@
+            self.update(X[i, :], Y[i, :])
+
+    def update(self, X, Y):
+        """Update model with a single observation.
+
+        When new input-output data is available, the model can be updated using
+        this method. As more and more data are collected, the number of
+        hypotheses grows, increasing the computational complexity. By default
+        hypotheses are pruned at the end of each update (see
+        :py:meth:`.trim`.). To disable hypotheses trimming, initialise the
+        class with ``maxhypot`` set to ``None``.
+
+        Args:
+            X (numpy.array): Observed (1 x M) input data (predictor variable).
+            Y (numpy.array): Observed (1 x N) output data (response variable).
+
+        """
+
+        # Initialise algorithm on first call to update. This allows the
+        # algorithm to configure itself to the size of the first input/output
+        # data if no hyper-parameters have been specified.
+        if not self.__initialised:
+            m = X.shape[1] if np.ndim(X) > 1 else X.size
+            n = Y.shape[1] if np.ndim(Y) > 1 else Y.size
+            self.__initialise_algorithm(m, n)
+            self.__initialised = True
+
+        # Get size of data.
+        k = X.shape[0] if np.ndim(X) > 1 else X.size
+        m, n = self.__m, self.__n
+
+        # Allocate variables for the dynamic programming pass.
+        loglik = -np.inf
+        logmax = -np.inf
+        logsum = -np.inf
+        ind = np.nan
+
+        # Update hypotheses by updating each maxtrix variate, normal inverse
+        # gamma distribution over the linear models.
+        self.__update_count += 1
+        for hypotheses in self.__hypotheses:
+
+            # Update the sufficient statistics.
+            hypotheses['distribution'].update(self.__featfun(X), Y)
+
+            # Compute the log-normalization constant after the update
+            # (posterior parameter distribution).
+            # (Equation 8)
+            n_o = hypotheses['log_constant']
+            n_k = hypotheses['log_constant'] = hypotheses['distribution'].logconst()
+
+            # Evaluate the log-density of the predictive distribution.
+            # (Equation 16)
+            log_density = n_k - n_o - k * (0.5 * m * n) * np.log(2.0 * np.pi)
+
+            # Increment the counter.
+            hypotheses['count'] += 1
+
+            # Accumulate the log-likelihood of the data.
+            # (Equation 17)
+            hazard = self.__ratefun(hypotheses['count'])
+            aux = np.log(hazard) + log_density + hypotheses['log_probability']
+            loglik = self.__accum(loglik, aux)
+
+            # Keep track of the highest, log-likelihood.
+            if aux > logmax:
+                logmax, ind = aux, hypotheses['count']
+
+            # Update and accumulate the log-probabilities.
+            hypotheses['log_probability'] += np.log1p(-hazard) + log_density
+            logsum = self.__accum(logsum, hypotheses['log_probability'])
```

Comment:
I didn't go over this in detail. I assume the equations haven't changed with respect to the original implementation.